### PR TITLE
Added Omens Quest (BLU AF2)

### PR DESCRIPTION
**Pull request by kaincenteno**
Changed BEGINNINGS to check correct AF1 global
Position of flans not accurate to retail as there's no captures for a full alliance fighting BCNM
renamed blank npc to Blu_Af2 so it's easier to tell what purpose it serves.
Immortal Flans do not have any special AI or move sets that differences them from other flans.
Thank you @Wiggo32  for capturing this quest.

**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](https://github.com/project-topaz/topaz/blob/master/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (KnowOne134) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: Just for readability of the code could you add local getqueststatus(..., omens) and then add that to the check with has envelope.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (KnowOne134) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: Can use the above local here also then

File: scripts/globals/bcnm.lua
Comment (KnowOne134) [inline scripts/globals/bcnm.lua]: Can we change the charger to the quest name. Quest_Omens for all scripts? Thoughts?

File: scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua
Comment (KnowOne134) [inline scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua:1]: No need for this script as it has no info at this time

File: sql/npc_list.sql
Comment (KnowOne134) [inline sql/npc_list.sql]: Is this the only thing this npc in the game does is deal with blu quest?

**Review by KnowOne134 — COMMENTED** [review]
Thanks for the work done. Have a few critiques before I actually test and to hear from others

File: sql/npc_list.sql
Comment (kaincenteno) [reply sql/npc_list.sql]: As far as i know it is only used for BLU AF2 quest. It's a blank npc on the floor. I can change it back but just named it before as on other PRs i submitted before i was asked to name it something related to what it is used for if its a blank or duplicate. (was just trying to avoid having to have discussions about it lol but again no issues changing it back just need confirmation :cat:  )

File: scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua
Comment (kaincenteno) [reply scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua:1]: Should I just delete it?  :scream_cat:

File: scripts/globals/bcnm.lua
Comment (kaincenteno) [reply scripts/globals/bcnm.lua]: Charger? what does that mean? :sweat_smile:

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: For sure but the check for envelope is only being used once. From old PRs I have submitted on the other project that shall not be named, I was told to do that only if it is being used more than once for readability :cat: 

Tagging @ibm2431  for second opinion.

File: sql/npc_list.sql
Comment (kaincenteno) [reply sql/npc_list.sql]: Tagging @TeoTwawki  for his insight in naming unnamed npcs :cat:

File: scripts/globals/bcnm.lua
Comment (KnowOne134) [reply scripts/globals/bcnm.lua]: Sorry. Phone and auto correct. 
CharVar is what that was meant to be

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (KnowOne134) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: Just the readability part. Like just looking at the code I wouldn't know it's for this quest because the check is just an envelope ki that I don't know where it comes from

File: scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua
Comment (KnowOne134) [reply scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua:1]: Git delete yes. Unless you want to add something into the lua for the mob

File: scripts/globals/bcnm.lua
Comment (kaincenteno) [reply scripts/globals/bcnm.lua]: Sure we can, that will help with readability :cat:
I'll change it today

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (ibm2431) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: General rule was that you could assign a variable if it makes an unreadably long line shorter, but...

I feel bad for `player:hasCompletedQuest()`, everyone always forgets it exists.

(edit: This is in reference to assigning a local to quest status to shorten lines when checking it against QUEST_COMPLETE)

File: sql/npc_list.sql
Comment (TeoTwawki) [reply sql/npc_list.sql]: it has no other uses, name doesn't much matter since blank in game except for being readable/recognizable to us humans.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: You're squishing your commas~!

(In lots of places)

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: Empty Vessel is the BLU unlock quest

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: Do me a favor and use `setLocalVar` and `getLocalVar` instead. I don't know if Waould is already using one if you've refused a divination (and if not, could you adjust that?), but `needToZone` should be phased out and eventually removed because it's "zone global" without a specified reason.

edit: Or, alternatively, be refactored to _take_ a reason and keep them straight.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: Does the event handle denying the player if they don't have enough gil? Or do you need to?

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua:190]: Either way, be sure to check the player's gil!

I think delGil won't take gil below 0, but there's still the message below, and it's good to keep the habit~

File: scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua
Comment (ibm2431) [inline scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua]: So question: Do additional flans spawn in front of an inside player's eyes as more players enter the battlefield? Should we be checking party numbers as players enter, or when the battlefield is initially set up?

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: Would you happen to know if Waoud offers a divination after speaking with Lathuya? This startEvent here is correct based off of Wiggo's capture, but he didn't speak with Waoud again between Lathuya and completion. (I wouldn't have either.)

File: sql/npc_list.sql
Comment (ibm2431) [reply sql/npc_list.sql]: Unless ships are involved. Don't ask me why.

(And I'd quip that "Omens" would be more clear than "Af2")

edit: This is in reference to names being whatever you want. Come on, GitHub.

**Review by ibm2431 — CHANGES_REQUESTED** [review]

File: sql/npc_list.sql
Comment (TeoTwawki) [reply sql/npc_list.sql]: > ships

coz moving zone parts.

also certain bcnm entrance (divine might..) and door etc, expect the name to match wtf they reference.

those are usually (nut not always named the hex value of whatever Nth position they exist in zone, prefixed with a symbol like: `_07f` or matching the fourcc of their accompanying animation loop for objects that just sit and sparkle but can't be clicked on (holiday decorations will animate automatically if correctly named).

/endClientTrivia

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (TeoTwawki) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: even if it does, never trust the event - injections a thing

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: Fixed. Will be pushing change soon.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: fixed. Will push change soon.

File: scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua
Comment (kaincenteno) [reply scripts/zones/Navukgo_Execution_Chamber/mobs/Immortal_Flan.lua:1]: Deleted. Pushing change soon.

File: sql/npc_list.sql
Comment (kaincenteno) [reply sql/npc_list.sql]: changing it back to 15, until there's a confirmation that this blank npc is not used for something else (can't be proven as most of Aht Ughan quests/misisions are not scripted).

Change will be seen on the next push (in a few minutes)

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: added spaces to commas through the file (most of the file was previously scripted by someone else)
Push coming soon

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: oh crap that's right. Thanks for mentioning it. Reverting this.
Push coming soon

Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: moved it to begginings, i knew somewhere else needed to be fixed lol

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: @ibm2431  The event handles the check if they dont have enough gil.
No changes done here

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua:190]: No need to check the player's Gil as the check is part of the params for it.

No changes made here.

File: scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua
Comment (kaincenteno) [reply scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua]: Flans are really far away from the players eyes, They cannot be seen from @Wiggo32 's capture. If the logic on this needs to be changed a capture with 4+ people would be needed to verify how do they pop. 

Let me know if any change is needed here :cat:

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: I'm unsure if it offers divination afterwards. No capture was done for this part.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: created localvar WaoudNeedToZone.

Pushing changes soon

File: sql/npc_list.sql
Comment (TeoTwawki) [reply sql/npc_list.sql]: I'm positive that it doesn't have any missions involved nor any non blu quest.

File: scripts/globals/bcnm.lua
Comment (kaincenteno) [reply scripts/globals/bcnm.lua]: resolved

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (KnowOne134) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: there is a function player:needToZone()

Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: ah thanks for the catch @KnowOne134 , let me murder it with kindness

Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: @KnowOne134 i dont see it on my end. You might be looking at a wrong commit?

Comment (TeoTwawki) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: just generally you don't want a var (in this case `local needToZone`) that has same name as a function. I recall ibm wanting to phase said function out.

Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: alright. Changing it @TeoTwawki  😸

Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua]: changed to waoudNeedToZone :smile_cat: 

resolving comment

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (ibm2431) [inline scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: I might prefer a:
```lua
if omens >= QUEST_ACCEPTED then
    if omensProgress == 3 then
        ...etc
    elseif omens == QUEST_COMPLETED then
        -- default dialogue
    end
end
```

instead of checking for quest progress by seeing if a character has a key item or not. In my view, key items are just shiny baubles that are given to you as a side effect of your actual progress (the variable) - not the progress themselves.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (ibm2431) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua:190]: Packet injection is a thing 😉 

When params are sent to the client in a `startEvent`, those params only tell a client how to change the display of the event. The resulting "option" that's sent by the client to the server is entirely under the client's control, and there's no way to prevent / guarantee a client can _only_ send certain options back to the server after being given specific params by us.

So instead, we need to be sure to _never trust the client_ and always re-verify that "meaningful" options they send (ie: resulting in giving items, or in this case, deleting gil) are actually valid and that we're not being lied to.

File: scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua
Comment (ibm2431) [reply scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua]: While waiting for other players to enter a BCNM, players have a tendency to run up to the edge of the mobs' aggro range.

Going to tag @KnowOne134 and @cocosolos here as they do more work with BCNMs than me - how would we go about setting the number of flans based on registrant party size? Set a "local-but-not-lua-local" var for number of flans needed during `onBattlefieldRegister`? Can we spawn all the required flans at once in `onBattlefieldInitialise`?

**Review by ibm2431 — CHANGES_REQUESTED** [review]

File: scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua
Comment (cocosolos) [reply scripts/zones/Navukgo_Execution_Chamber/bcnms/omens.lua]: If you wanna use party/alliance size you can use onBattlefieldInitialise and get the party size with ```battlefield:getInitiator():getPartySize()``` or ```battlefield:getInitiator():getAllianceSize()```. Players do get locked out once one of the mobs is engaged so onBattlefieldEnter works as well, depends on how you want the mobs to spawn. For example if someone does an early pull before everyone enters, should they still be fighting the full amount of mobs?

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua:190]: ah got it, but in this case it's deleting gil so I would assume if someone is trying to cheat then they must be really dumb to cheat to delete their items lol.

If we are trying to avoid cheating then it would be for other stuff like unlocking this job without meeting the criteria.

I can go and change this and make all the onEventFinish to have the same requirements that started that event :cat: let me know.

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Lathuya.lua]: looks nicer :+1:  thank you @ibm2431

File: scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua
Comment (kaincenteno) [reply scripts/zones/Aht_Urhgan_Whitegate/npcs/Waoud.lua:190]: after thinking it, you are right @ibm2431 :cat:  it makes sense to not be fooled by those cheaters that like to mess the game experience.

I only included it inside the OMENS onEventFinish. But did it fully on Lathuya. Maybe it could be a good project to be placed on the projects section :sunglasses:

**Review by ibm2431 — APPROVED** [review]
Pushed a commit to make the flan spawning behavior accurate to retail (and some miscellaneous fixes and changes).

Approved now, sorry it took me so long to get around to this, kain!

I feel weird merging my own code without a checkmark from _someone_, so I'm holding off on merge until someone-other-than-me says it looks good (or next staff meeting, whichever comes first).

**Pull request merged**
